RecoLocalMuon/GEMSegment/plugins/GE0SegAlgoRU.cc

```diff
@@ -0,0 +1,579 @@
+    });
+  else
+    std::sort(hitAndPositions.begin(), hitAndPositions.end(), [](const HitAndPosition& h1, const HitAndPosition& h2) {
+      return h1.layer > h2.layer;
+    });
+  return run(ensemble.first, hitAndPositions);
+}
+
+std::vector<GEMSegment> GE0SegAlgoRU::run(const GEMSuperChamber* chamber, const HitAndPositionContainer& rechits) {
+#ifdef EDM_ML_DEBUG  // have lines below only compiled when in debug mode
+  GEMDetId chId(chamber->id());
+  edm::LogVerbatim("GE0SegAlgoRU") << "[GEMSegmentAlgorithm::run] build segments in chamber " << chId
+                                   << " which contains " << rechits.size() << " rechits";
+  for (unsigned int iH = 0; iH < rechits.size(); ++iH) {
+    auto ge0id = rechits[iH].rh->gemId();
+    auto rhLP = rechits[iH].lp;
+    edm::LogVerbatim("GE0SegAlgoRU") << "[RecHit :: Loc x = " << std::showpos << std::setw(9) << rhLP.x()
+                                     << " Glb y = " << std::showpos << std::setw(9) << rhLP.y()
+                                     << " Time = " << std::showpos << rechits[iH].rh->BunchX() << " -- "
+                                     << ge0id.rawId() << " = " << ge0id << " ]" << std::endl;
+  }
+#endif
+
+  if (rechits.size() < 3 || rechits.size() > 300) {
+    return std::vector<GEMSegment>();
+  }
+
+  theChamber = chamber;
+
+  std::vector<unsigned int> recHits_per_layer(6, 0);
+  for (unsigned int iH = 0; iH < rechits.size(); ++iH) {
+    recHits_per_layer[rechits[iH].layer - 1]++;
+  }
+
+  BoolContainer used(rechits.size(), false);
+
+  // We have at least 2 hits. We intend to try all possible pairs of hits to start
+  // segment building. 'All possible' means each hit lies on different layers in the chamber.
+  // after all same size segs are build we get rid of the overcrossed segments using the chi2 criteria
+  // the hits from the segs that are left are marked as used and are not added to segs in future iterations
+  // the hits from 3p segs are marked as used separately in order to try to assamble them in longer segments
+  // in case there is a second pass
+
+  // Choose first hit (as close to IP as possible) h1 and second hit
+  // (as far from IP as possible) h2 To do this we iterate over hits
+  // in the chamber by layer - pick two layers.  Then we
+  // iterate over hits within each of these layers and pick h1 and h2
+  // these.  If they are 'close enough' we build an empty
+  // segment.  Then try adding hits to this segment.
+
+  std::vector<GEMSegment> segments;
+
+  auto doStd = [&]() {
+    for (unsigned int n_seg_min = 6u; n_seg_min >= stdParameters.minNumberOfHits; --n_seg_min)
+      lookForSegments(stdParameters, n_seg_min, rechits, recHits_per_layer, used, segments);
+  };
+  auto doDisplaced = [&]() {
+    for (unsigned int n_seg_min = 6u; n_seg_min >= displacedParameters.minNumberOfHits; --n_seg_min)
+      lookForSegments(displacedParameters, n_seg_min, rechits, recHits_per_layer, used, segments);
+  };
+  // Not currently used
```

Comment:
@kpedro88 we will have a student check the GE0 vs ME0 setup soon, and we are also planning to have them test the various options after the GE0 validation. There is a wide switch, but the code as written would switch various paths that I would prefer to just keep off until we understand all the effects it would have.